src/simplify_reshapes.cpp

```diff
@@ -318,11 +318,209 @@ struct find_nested_concat
+        {
+            if(scales[ii] == 1 or in_dims[ii] == 1)
+            {
+                continue;
+            }
+
+            out_dims[ii] = in_dims[ii];
+            in_dims.insert(in_dims.begin() + ii + 1, 1);
+            out_dims.insert(out_dims.begin() + ii + 1, scales[ii]);
+        }
+
+        auto in_rsp   = ins_rsp->inputs().front();
+        auto rsp_data = p.insert_instruction(
+            ins_rsp, migraphx::make_op("reshape", {{"dims", in_dims}}), in_rsp);
+        auto mb_rsp = p.insert_instruction(
+            ins_rsp, migraphx::make_op("multibroadcast", {{"output_lens", out_dims}}), rsp_data);
+        auto std_mb = p.insert_instruction(ins, migraphx::make_op("contiguous"), mb_rsp);
+        std::vector<int64_t> rsp_dims(out_lens.begin(), out_lens.end());
+        p.replace_instruction(ins, migraphx::make_op("reshape", {{"dims", rsp_dims}}), std_mb);
+    }
+};
+
+struct find_where_op
+{
+    auto matcher() const
+    {
+        return match::name("gather")(
+            match::args(match::name("reshape")(match::arg(0)(match::name("concat").bind("data"))),
+                        match::is_constant().bind("ind")));
+    }
+
+    void apply(module& p, match::matcher_result r) const
+    {
+        auto ins     = r.result;
+        auto concat  = r.instructions["data"];
+        auto ins_ind = r.instructions["ind"];
+        // if ind is not constant, cannot optimize
+        std::vector<bool> vec_ind;
+        auto arg_ind = ins_ind->eval();
+        if(arg_ind.empty())
+        {
+            return;
+        }
+        arg_ind.visit([&](auto v) { vec_ind.assign(v.begin(), v.end()); });
+        // ind has to be the same value
+        auto val = vec_ind.front();
+        if(not std::all_of(vec_ind.begin(), vec_ind.end(), [&](auto v) { return (v == val); }))
+        {
+            return;
+        }
+
+        // concat axis must be 0
+        auto op = any_cast<op::concat>(concat->get_operator());
+        if(op.axis != 0)
+        {
+            return;
+        }
+
+        // check concat inputs, it has to be 2 and have the same shape
+        const auto& inputs = concat->inputs();
+        if(inputs.size() != 2)
+        {
+            return;
+        }
+        if(inputs.at(0)->get_shape() != inputs.at(1)->get_shape())
+        {
+            return;
+        }
+        if(inputs.at(0)->get_shape().lens() != ins_ind->get_shape().lens())
+        {
+            return;
+        }
+
+        if(val)
+        {
+            p.replace_instruction(ins, inputs.at(0));
+        }
+        else
+        {
+            p.replace_instruction(ins, inputs.at(1));
+        }
+    }
+};
+
+struct find_reshape_cont
+{
+    auto matcher() const
+    {
+        return match::name("reshape")(match::args(match::name("contiguous").bind("cont")),
+                                      match::used_once(),
+                                      match::any_of[match::outputs()](match::pointwise()));
```

Comment:
Instead of matching the outputs, it should match the input to pointwise is a reshape, something like this:

```cpp
    auto matcher() const
    {
        return match::pointwise(match::either_arg(0, 1)(match::name("reshape")(
            match::args(match::name("contiguous").bind("cont")),
            match::used_once())));
    }
```